psi4/driver/procrouting/scf_proc/smart_scf.py

```diff
@@ -0,0 +1,168 @@
+        self.Drms_history = []
+        self.user_vals = {
+            "frac_enabled": self.wfn.frac_enabled,
+            "damping_enabled": self.wfn.damping_enabled,
+            "soscf_enabled": self.wfn.soscf_enabled,
+            "MAXITER": core.get_option('SCF', 'MAXITER')
+        }
+
+        #self.stage is where options are directly set by decision methods.
+        self.stage = {}
+
+    def smart_iter(self, SCFE, Drms):
+        """Called every iteration to update energy history,
+        density history, and call decision making methods.
+        
+        Decision making methods, such as initdamp(), edit a dictionary
+        called self.smart_opts_dict. These are 'staged' options, which are
+        then validated for conflicts at the end of this method. 
+        """
+
+        #Update Energy and Drms history
+        self.update_E_history()
+        self.update_Drms_history(Drms)
+        if not self.initdamp():
+            self.trailing_conv()
+
+        #Finally, check for conflicts between set options
+        _validate_smart()
+
+    def trailing_conv(self):
+        """Detects trailing-convergence and attempts to alleviate the issue
+        by engaging SOSCF measures and a conservative amount of damping.
+
+        The criteria for 'trailing convergence' are:
+            Greater than 15 SCF cycles have occured,
+            the most recent Drms value is less than 1E-4,
+            the requested convergence is tighter (lower in magnitude) than
+                1E-8
+        
+        These criteria are somewhat arbitrary but seem reasonable to MMD.
+
+        """
+
+        if (self.wfn.iteration_ > 15 and core.get_options('SCF', 'E_CONVERGENCE') < 1.0E-8
+                and self.Drms_history[-1] <= 1E-4):
+            self.wfn.soscf_enabled = True
+            self.wfn.damping_enabled = True
+            self.wfn.damping_percentage = 25
+            return True
+
+        else:
+            return False
+
+    def initdamp(self):
+        """Decides whether to damp initial iterations in SCF.
+
+        The criteria are:
+            SCF iteration < 4, 
+            and ither the SAD, GWH, or CORE guess was selected.
+        """
+
+        #If the iteration is not <= 4, no need to continue.
+        if self.wfn.iteration_ > 4:
+            return False
+
+        guess_opt = core.get_option('SCF', "GUESS")
+
+        if (guess_opt in {'SAD', 'GWH', 'CORE'}):
+            self.stage['DAMPING_PERCENTAGE'] = smart_vals['INIT_DAMP_PERCENT']
+            self.stage['DAMPING_ENABLED'] = True
+            return True
+
+    def dyn_damp(self, Drms_target=5E-4):
+        """
+        Offers dynamic damping to hit a target Drms value.
+        Default is 5E-4
+        """
+
+        if self.wfn.iterations_ < 5:
+            return 0
+        elif self.Drms_history[-1] > Drms_target:
+            self.stage['DAMPING_PERCENTAGE'] = \
+                    Drms_target/self.Drms_history[-1]
+
+    def update_E_history(self):
```

Comment:
Seems like you never want to call `update_*` separately, best not to have these as functions and insert them in the `smart_iter` section.